### PR TITLE
Merge Personal/yufei/extend type rules to master

### PR DESCRIPTION
**Pull request by moria97**
1. Upgrade configuration file to support extended type rules like "HumanName.family":"redact",
 "Reference.display":"redact"
2. Enable path and type validation by constructing FHIR Schema from Hl7.Fhir package

### CODE REVIEW

File: src/Fhir.Anonymizer.Core/AnonymizationConfigurations/AnonymizerRule.cs [hunk 1]
```diff
@@ -9,13 +9,16 @@ public class AnonymizerRule
         public AnonymizerRuleType Type { get; set; }
 
         public string Source { get; set; }
+        // Rule priority, based on line sequence
```
Comment (deepak-ms) [inline src/Fhir.Anonymizer.Core/AnonymizationConfigurations/AnonymizerRule.cs]: Will we have separate path and type rules even with priority defined by line sequence?

Comment (moria97) [reply src/Fhir.Anonymizer.Core/AnonymizationConfigurations/AnonymizerRule.cs]: For now, we prefer to keep seperately as paths in path rules are compatiable with FHIR paths and the type rules are not.  And we might need to implement wildcard and conditional anonymization taking advantage of FHIRPath compiler which is powerful with functions, so we'd better leave them seperately for future compatibility as there will be incremental changes to the configuration.
We will merge these rules finally when the configuration file comes to a steady state.

File: src/Fhir.Anonymizer.Core/AnonymizationConfigurations/ResourceAnonymizerContext.cs [hunk 2]
```diff
@@ -7,55 +7,56 @@ namespace Fhir.Anonymizer.Core.AnonymizerConfigurations
...
+            var currentNode = node;
+            var typePath = string.Empty;
+            do
             {
-                TransformTypeRulesToPathRules(child, typeRules, rules, rulePaths, generatedTypePaths);
-            }
+                var path = string.IsNullOrEmpty(typePath) ? currentNode.InstanceType : $"{currentNode.InstanceType}.{typePath}";
+                var ruleForPath = _typeRuleMap.GetValueOrDefault(path, null);
+
+                if (ruleForPath != null && ruleForPath.Priority < maxPriority)
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Core/AnonymizationConfigurations/ResourceAnonymizerContext.cs]: At this point, the smaller the ruleForPath.Priority is, the higher its priority is, right? The naming of maxPriority looks that it is the highest priority. Could we change maxPriority to lowestPriority or something like that?

File: src/Fhir.Anonymizer.Core/AnonymizationConfigurations/ResourceAnonymizerContext.cs [hunk 3]
```diff
@@ -7,55 +7,56 @@ namespace Fhir.Anonymizer.Core.AnonymizerConfigurations
...
-            var children = node.Children().Cast<ElementNode>();
-            foreach (var child in children)
+            var currentNode = node;
+            var typePath = string.Empty;
+            do
             {
-                TransformTypeRulesToPathRules(child, typeRules, rules, rulePaths, generatedTypePaths);
-            }
+                var path = string.IsNullOrEmpty(typePath) ? currentNode.InstanceType : $"{currentNode.InstanceType}.{typePath}";
+                var ruleForPath = _typeRuleMap.GetValueOrDefault(path, null);
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Core/AnonymizationConfigurations/ResourceAnonymizerContext.cs]: The naming of variable path here seems a bit strange, as it is passed as a parameter into a type rule map.

File: src/Fhir.Anonymizer.Core/AnonymizationConfigurations/Validation/AnonymizerConfigurationValidator.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            var invalidPathRuleCount = 0;
+            if (config.PathRules != null)
+            {
+                foreach (var rule in config.PathRules)
+                {
+                    var validationResult = _fhirSchemaProvider.ValidateRule(rule.Key, rule.Value, AnonymizerRuleType.PathRule, _anonymizationMethodTargetTypes.GetValueOrDefault(rule.Value));
+                    if (!validationResult.Success)
+                    {
+                        invalidPathRuleCount++;
+                        Console.Error.WriteLine($"Validate rule [{rule.Key}:{rule.Value}] failed: {validationResult.ErrorMessage}");
```
Comment (BoyaWu10) [inline src/Fhir.Anonymizer.Core/AnonymizationConfigurations/Validation/AnonymizerConfigurationValidator.cs]: Do we use _logger.LogError() here? So all information in core can be produced with logger.

**Pull request closed without merging**